pkg/devicetwin/dtcontroller_test.go

```diff
@@ -0,0 +1,757 @@
+				Stop:              make(chan bool, 1),
+			},
+			wantError: nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := InitDTController(test.context)
+			if !reflect.DeepEqual(test.wantError, err) {
+				t.Errorf("InitDTController() error = %v, wantError %v", err, test.wantError)
+				return
+			}
+			if reflect.TypeOf(got) != reflect.TypeOf(test.want) {
+				t.Errorf("InitDTController() = %v, wantError %v", got.DTContexts, test.want.DTContexts)
+				return
+			}
+			if !reflect.DeepEqual(got.HeartBeatToModule, test.want.HeartBeatToModule) {
+				t.Errorf("InitDTController() failed due to wrong HeartBeatToModule, Got= %v Want = %v", got.HeartBeatToModule, test.want.HeartBeatToModule)
+				return
+			}
+			if !reflect.DeepEqual(got.DTContexts.ModulesContext, test.want.DTContexts.ModulesContext) {
+				t.Errorf("IniDTController() failed due to wrong context, Got =%v Want = %v", got.DTContexts.ModulesContext, test.want.DTContexts.ModulesContext)
+				return
+			}
+			if !reflect.DeepEqual(got.DTModules, test.want.DTModules) {
+				t.Errorf("InitDTController() failed due to wrong DTModules, Got =%v Want =%v", got.DTModules, test.want.DTModules)
+				return
+			}
+			if cap(got.Stop) != cap(test.want.Stop) {
+				t.Errorf("InitDTController failed due to wrong Stop Chan Size,Got = %v Want =%v", got.Stop, test.want.Stop)
+			}
+		})
+	}
+}
+
+//TestRegisterDTModule is function to test RegisterDTmodule().
+func TestRegisterDTModule(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContexts, _ := dtcontext.InitDTContext(mainContext)
+	dtc := &DTController{
+		HeartBeatToModule: make(map[string]chan interface{}),
+		DTContexts:        dtContexts,
+		DTModules:         make(map[string]dtmodule.DTModule),
+		Stop:              make(chan bool, 1),
+	}
+	tests := []struct {
+		name       string
+		modulename string
+	}{
+		{
+			name:       "MemModule",
+			modulename: "MemModule",
+		},
+		{
+			name:       "TwinModule",
+			modulename: "TwinModule",
+		},
+		{
+			name:       "CommModule",
+			modulename: "CommModule",
+		},
+		{
+			name:       "DeviceModule",
+			modulename: "DeviceModule",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			dtc.RegisterDTModule(tt.modulename)
+			test := false
+			for _, name := range dtc.DTModules {
+				test = false
+				if name.Name == tt.modulename {
+					test = true
+					break
+				}
+			}
+			if test == false {
+				t.Errorf("RegisterDTModule failed to register the module %v", tt.modulename)
+			}
+		})
+	}
+}
+
+//TestDTontroller_Start is function to test Start().
+func TestDTController_Start(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContexts, _ := dtcontext.InitDTContext(mainContext)
+	mainContext.AddModule("twin")
+	// fakeDevice is used to set the argument of All function
+	fakeDevice := new([]dtclient.Device)
```

Comment:
the code to create fake devices , attributes, and twins is repeated many places in different test cases. its better we can have common code which is resued by all tests

Reply:
Updated